lib/assert.js

```diff
@@ -140,95 +140,151 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
   }
 };
 
+function areSimilarRegExps(a, b) {
+  return a.source === b.source && a.flags === b.flags;
+}
+
+function areSimilarTypedArrays(a, b) {
+  return compare(Buffer.from(a.buffer,
+                             a.byteOffset,
+                             a.byteLength),
+                 Buffer.from(b.buffer,
+                             b.byteOffset,
+                             b.byteLength)) === 0;
+}
+
+function isNullOrNonObj(object) {
+  return object === null || typeof object !== 'object';
+}
+
+function isFloatTypedArray(object) {
+  const tag = getTag(object);
+  return tag === '[object Float32Array]' || tag === '[object Float64Array]';
+}
+
 function _deepEqual(actual, expected, strict, memos) {
   // All identical values are equivalent, as determined by ===.
   if (actual === expected) {
     return true;
+  }
 
-  // If both values are instances of buffers, equivalence is
-  // determined by comparing the values and ensuring the result
-  // === 0.
-  } else if (actual instanceof Buffer && expected instanceof Buffer) {
-    return compare(actual, expected) === 0;
+  // Do fast checks for builtin types.
+  // If they don't match, they must not be equal.
+  // If they match, return true for non-strict comparison.
+  // For strict comparison we need to exam further.
 
-  // If the expected value is a Date object, the actual value is
-  // equivalent if it is also a Date object that refers to the same time.
-  } else if (util.isDate(actual) && util.isDate(expected)) {
-    return actual.getTime() === expected.getTime();
+  // If both values are instances of Buffers,
+  // check if compare returns 0 first.
+  if (actual instanceof Buffer && expected instanceof Buffer) {
+    if (compare(actual, expected) !== 0) {
+      return false;
+    } else if (!strict) {
+      return true;  // Skip further checks for non-strict comparison.
+    }
+  }
```

Comment:
This would avoid two extra `Buffer.from` calls when both of them are Buffers(might be expensive). I think it is worth keeping, unless we change how TypedArrays are compared(e.g. just loop over the elements).

No sure why the TypedArray check is implemented that way in the first place though, if it's for performance then wouldn't the overhead of calling into c++ and the copying in `Buffer.from` just kill that?